controllers/vsl.go

```diff
@@ -261,3 +266,33 @@ func containsPlugin(d []oadpv1alpha1.DefaultPlugin, value string) bool {
 	}
 	return false
 }
+
+func (r *DPAReconciler) ensureVslSecretDataExists(dpa *oadpv1alpha1.DataProtectionApplication, vsl *oadpv1alpha1.SnapshotLocation) error {
```

Comment:
this is a duplication of  `ensureSecretDataExists` function in `controllers/bsl.go`, but I was not able to remove it, compiler always complained about my ideas. Is it possible to have a single function for both bsl and vsl?

Reply:
I also tried it actually, but there are some differences between bsl and vsl specs, so it was making it more complex,  also considering the future use , I thought keeping it separate will be better.

Do you have some other ideas? @mateusoliveira43

Reply:
How about something like this, it compiles but didn't test in the OCP with reconcile nor e2e tests, just an idea:

```diff
diff --git a/controllers/bsl.go b/controllers/bsl.go
index 005b851..2140184 100644
--- a/controllers/bsl.go
+++ b/controllers/bsl.go
@@ -403,29 +403,9 @@ func (r *DPAReconciler) ensurePrefixWhenBackupImages(dpa *oadpv1alpha1.DataProte
 }
 
 func (r *DPAReconciler) ensureSecretDataExists(dpa *oadpv1alpha1.DataProtectionApplication, bsl *oadpv1alpha1.BackupLocation) error {
-       // Check if the Velero feature flag 'no-secret' is not set
-       if !(dpa.Spec.Configuration.Velero.HasFeatureFlag("no-secret")) {
-               // Check if the user specified credential under velero
-               if bsl.Velero != nil && bsl.Velero.Credential != nil {
-                       // Check if user specified empty credential key
-                       if bsl.Velero.Credential.Key == "" {
-                               return fmt.Errorf("Secret key specified in BackupLocation %s cannot be empty", bsl.Name)
-                       }
-                       // Check if user specified empty credential name
-                       if bsl.Velero.Credential.Name == "" {
-                               return fmt.Errorf("Secret name specified in BackupLocation %s cannot be empty", bsl.Name)
-                       }
-               }
-               // Check if the BSL secret key configured in the DPA exists with a secret data
-               secretName, secretKey := r.getSecretNameAndKeyforBackupLocation(*bsl)
-               bslSecret, err := r.getProviderSecret(secretName)
-               if err != nil {
-                       return err
-               }
-               data, foundKey := bslSecret.Data[secretKey]
-               if !foundKey || len(data) == 0 {
-                       return fmt.Errorf("Secret name %s is missing data for key %s", secretName, secretKey)
-               }
+       getLocationNameAndKeyFunc := func(loc interface{}) (string, string) {
+               bsl := loc.(*oadpv1alpha1.BackupLocation)
+               return r.getSecretNameAndKeyforBackupLocation(*bsl)
        }
-       return nil
+       return r.ensureSecretDataExistsCommon(dpa, bsl, getLocationNameAndKeyFunc)
 }
diff --git a/controllers/common.go b/controllers/common.go
index 80f5eeb..77aee99 100644
--- a/controllers/common.go
+++ b/controllers/common.go
@@ -1,11 +1,54 @@
 package controllers
 
 import (
+       "fmt"
+       oadpv1alpha1 "github.com/openshift/oadp-operator/api/v1alpha1"
        "github.com/openshift/oadp-operator/pkg/common"
        corev1 "k8s.io/api/core/v1"
        "k8s.io/utils/pointer"
+       "reflect"
 )
 
+func (r *DPAReconciler) ensureSecretDataExistsCommon(dpa *oadpv1alpha1.DataProtectionApplication, location interface{}, getLocationNameAndKeyFunc func(interface{}) (string, string)) error {
+       // Check if the Velero feature flag 'no-secret' is not set
+       if !(dpa.Spec.Configuration.Velero.HasFeatureFlag("no-secret")) {
+               // Check if the user specified credential under velero
+               var cred *corev1.SecretKeySelector
+               var name string
+               switch loc := location.(type) {
+               case *oadpv1alpha1.BackupLocation:
+                       cred = loc.Velero.Credential
+                       name = cred.Name
+               case *oadpv1alpha1.SnapshotLocation:
+                       cred = loc.Velero.Credential
+                       name = cred.Name
+               default:
+                       return fmt.Errorf("unknown location type")
+               }
+               if cred != nil {
+                       // Check if user specified empty credential key
+                       if cred.Key == "" {
+                               return fmt.Errorf("Secret key specified in %s %s cannot be empty", reflect.TypeOf(location).Elem().Name(), name)
+                       }
+                       // Check if user specified empty credential name
+                       if name == "" {
+                               return fmt.Errorf("Secret name specified in %s %s cannot be empty", reflect.TypeOf(location).Elem().Name(), name)
+                       }
+               }
+               // Check if the secret key configured in the DPA exists with a secret data
+               secretName, secretKey := getLocationNameAndKeyFunc(location)
+               locationSecret, err := r.getProviderSecret(secretName)
+               if err != nil {
+                       return err
+               }
+               data, foundKey := locationSecret.Data[secretKey]
+               if !foundKey || len(data) == 0 {
+                       return fmt.Errorf("Secret name %s is missing data for key %s", secretName, secretKey)
+               }
+       }
+       return nil
+}
+
 // setting defaults to avoid emitting update events
 func setContainerDefaults(container *corev1.Container) {
        // setting defaults to avoid emitting update events
diff --git a/controllers/vsl.go b/controllers/vsl.go
index 7e4cdb2..ad7efc6 100644
--- a/controllers/vsl.go
+++ b/controllers/vsl.go
@@ -268,31 +268,9 @@ func containsPlugin(d []oadpv1alpha1.DefaultPlugin, value string) bool {
 }
 
 func (r *DPAReconciler) ensureVslSecretDataExists(dpa *oadpv1alpha1.DataProtectionApplication, vsl *oadpv1alpha1.SnapshotLocation) error {
-       // Check if the Velero feature flag 'no-secret' is not set
-       if !(dpa.Spec.Configuration.Velero.HasFeatureFlag("no-secret")) {
-               // Check if the user specified credential under velero
-               if vsl.Velero != nil && vsl.Velero.Credential != nil {
-                       // Check if user specified empty credential key
-                       if vsl.Velero.Credential.Key == "" {
-                               return fmt.Errorf("Secret key specified in SnapshotLocation %s cannot be empty", "SL")
-                       }
-                       // Check if user specified empty credential name
-                       if vsl.Velero.Credential.Name == "" {
-                               return fmt.Errorf("Secret name specified in SnapshotLocation %s cannot be empty", "SL")
-                       }
-
-               }
-               // Check if the VSL secret key configured in the DPA exists with a secret data
-               secretName, secretKey := r.getSecretNameAndKeyforSnapshotLocation(*vsl)
-               vslSecret, err := r.getProviderSecret(secretName)
-               if err != nil {
-                       return err
-               }
-               data, foundKey := vslSecret.Data[secretKey]
-               if !foundKey || len(data) == 0 {
-                       return fmt.Errorf("Secret name %s is missing data for key %s", secretName, secretKey)
-               }
-
+       getLocationNameAndKeyFunc := func(loc interface{}) (string, string) {
+               vsl := loc.(*oadpv1alpha1.SnapshotLocation)
+               return r.getSecretNameAndKeyforSnapshotLocation(*vsl)
        }
-       return nil
+       return r.ensureSecretDataExistsCommon(dpa, vsl, getLocationNameAndKeyFunc)
 }

```

Reply:
yes i tried exactly like this typecasting approach, using interface{}, maybe it will work, but dont you think it is more complicated approach?

Reply:
i couldnt get it working though, using this as well.

Reply:
AFAIR, one of the problem i faced is comparing "loc.Velero!=nil" inside the common function (which is missing in above code) as bsl.velero and vsl.velero are of different types BackupLocationSpec, and Volumesnapshotlocationspec.

Reply:
It's definitely less readable code. Other option would be to have very simple functs that does just parts of the common logic, e.g.:
```
if !(dpa.Spec.Configuration.Velero.HasFeatureFlag("no-secret"))
```
Or the following to write as fucnt getting secretName and secretKey as args
```
		if !foundKey || len(data) == 0 {
			return fmt.Errorf("Secret name %s is missing data for key %s", secretName, secretKey)
		}
```

Reply:
@mpryc yes, this would be better and much readable, Updated! Please have a look.

Reply:
1 function for getting Secret name and key, and 1 function for validation.